File: Data Structures & Algorithms/max-area-of-island/submission-0.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        row, col = len(grid), len(grid[0])
        vis = set()

        def dfs(i, j):
            if (
                i < 0 or i >= row or
                j < 0 or j >= col or
                grid[i][j] == 0 or
                (i, j) in vis
            ):
                return 0

            vis.add((i, j))
            area = 1
            area += dfs(i + 1, j)
            area += dfs(i - 1, j)
            area += dfs(i, j + 1)
            area += dfs(i, j - 1)
            return area

        ans = 0
        for i in range(row):
            for j in range(col):
                if grid[i][j] == 1 and (i, j) not in vis:
                    ans = max(ans, dfs(i, j))

        return ans

File: Data Structures & Algorithms/count-connected-components/submission-0.cpp
class Solution {
public:

    void dfs(int i,vector<bool> &vis,vector<vector<int>>& gr){
        vis[i]=true;
        for(int j:gr[i]){
            if(!vis[j]){
                dfs(j,vis,gr);
            }
        }
    }
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> gr(n,vector<int> ());

        for(auto e:edges){
            gr[e[0]].push_back(e[1]);
            gr[e[1]].push_back(e[0]);
        }

        int cnt=0;

        vector<bool> vis(n,false);
        for(int i=0;i<n;i++){
            if(!vis[i]){
                dfs(i,vis,gr);
                cnt++;
            }
        }

        return cnt;

    }
};
